### PR TITLE
Segmind SD Distillation Blog

### PR DESCRIPTION
**Pull request by Warlord-K**
Added In Images and Blog Markdown for the Open Sourcing of Distilled Stable Diffusion Models by Segmind, The Blog highlights the need for distillation, the procedure of the distillation and the final results achieved along with the links to the Github Repo for the distillation code and links to the checkpoints.

### CODE REVIEW

File: sd_distillation.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# Open-sourcing Knowledge Distillation Code and Weights of SD-Small and SD-Tiny
```
Comment (sayakpaul) [inline sd_distillation.md:1]: We don't keep the assets in this repository except for the thumbnail. We keep the assets here: https://huggingface.co/datasets/huggingface/documentation-images. Feel free to submit a PR there.

File: sd_distillation.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
+# Open-sourcing Knowledge Distillation Code and Weights of SD-Small and SD-Tiny
+
+![image](assets/distill_sd/Picture1.png)
```
Comment (sayakpaul) [inline sd_distillation.md:3]: Let's follow the blog metadata format. You can review any of the raw .md files in this repo to get an idea. Also, we need to add an entry for this blog post to `_blogs.yml`.

File: sd_distillation.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+We, at Segmind, have been working on how to make the models faster and cheaper. Last year, we have open-sourced our accelerated SD-WebUI library called [voltaML](https://github.com/VoltaML/voltaML-fast-stable-diffusion), which is a AITemplate/TensorRT based inference acceleration library that has delivered between 4-6X increase in the inference speed. To continue towards the goal of making generative models faster, smaller and cheaper, we are open-sourcing the weights and training code of our compressed **SD models; SD-Small and SD-Tiny**. The pretrained checkpoints are available on [Huggingface 🤗](https://huggingface.co/segmind)
```
Comment (sayakpaul) [inline sd_distillation.md:8]: Awesome intro 🔥

File: sd_distillation.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+We, at Segmind, have been working on how to make the models faster and cheaper. Last year, we have open-sourced our accelerated SD-WebUI library called [voltaML](https://github.com/VoltaML/voltaML-fast-stable-diffusion), which is a AITemplate/TensorRT based inference acceleration library that has delivered between 4-6X increase in the inference speed. To continue towards the goal of making generative models faster, smaller and cheaper, we are open-sourcing the weights and training code of our compressed **SD models; SD-Small and SD-Tiny**. The pretrained checkpoints are available on [Huggingface 🤗](https://huggingface.co/segmind)
+
+## Knowledge Distillation:
```
Comment (sayakpaul) [inline sd_distillation.md:10]:
```suggestion
## Knowledge Distillation
```

File: sd_distillation.md (hunk 3 above)
Comment (sayakpaul) [reply sd_distillation.md:8]: Let's also add a table of contents as well?

File: sd_distillation.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+## Knowledge Distillation:
+
+![image](assets/distill_sd/Picture2.png)
+
+Our new compressed models have been trained on Knowledge-Distillation (KD) techniques and the work has been largely based on this paper. The authors describe a Block-removal Knowledge-Distillation method where some of the UNet layers are removed and the student model weights are trained. Using the KD methods described in the paper, we were able to train two compressed models using the Huggingface 🤗 [Diffusers 🧨](https://github.com/huggingface/diffusers) library ; Small and Tiny, that have 35% and 55% fewer parameters respectively than the base model while achieving comparable image fidelity as the base model. We have open sourced our distillation code in this [repo](https://github.com/segmind/distill-sd) and pretrained checkpoints on [Huggingface 🤗](https://huggingface.co/segmind)
```
Comment (sayakpaul) [inline sd_distillation.md:14]:
```suggestion
Our new compressed models have been trained on Knowledge-Distillation (KD) techniques and the work has been largely based on [this paper](https://openreview.net/forum?id=bOVydU0XKC). The authors describe a Block-removal Knowledge-Distillation method where some of the UNet layers are removed and the student model weights are trained. Using the KD methods described in the paper, we were able to train two compressed models using the [🧨 diffusers](https://github.com/huggingface/diffusers) library; **Small** and **Tiny**, that have 35% and 55% fewer parameters, respectively than the base model while achieving comparable image fidelity as the base model. We have open-sourced our distillation code in this [repo](https://github.com/segmind/distill-sd) and pretrained checkpoints on [Huggingface 🤗](https://huggingface.co/segmind).
```

The sentences here read arbitrarily long. Could we try to break them in smaller pieces?

File: sd_distillation.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+Combining all of this makes up the Knowledge-Distillation training. Below is an architecture of the Block Removed UNet used in the KD as described in the paper.
+
+![image](assets/distill_sd/Picture3.png)
+
+Image taken from the [paper](https://arxiv.org/pdf/2305.15798.pdf)  “On Architectural Compression of Text-to-Image Diffusion Models” by Shinkook. et. al
```
Comment (sayakpaul) [inline sd_distillation.md:24]:
```suggestion
Image taken from the [paper](https://arxiv.org/abs/2305.15798)  “On Architectural Compression of Text-to-Image Diffusion Models” by Shinkook. et. al.
```

File: sd_distillation.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+
+We have taken Realistic-Vision 4.0 as our base teacher model and have trained on the LAION Art Aesthetic dataset with image scores above 7.5, because of their high quality image descriptions. Unlike the paper, we have chosen to train the two models on 1M images for 100K steps for the Small and 125K steps for the Tiny mode respectively. The code for the distillation training can be found [here](https://github.com/segmind/distill-sd).
```
Comment (sayakpaul) [inline sd_distillation.md:27]: Provide links to:

* Realistic-Vision 4.0
* LAION Art Aesthetic dataset

File: sd_distillation.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+## Speed in terms of inference latency:
+
+We have observed that distilled models are up to 100% faster than the original base models.
+
+![image](assets/distill_sd/Picture4.png)
```
Comment (sayakpaul) [inline sd_distillation.md:33]: Where can the readers refer to for knowing more details about how this benchmarking was conducted?

File: sd_distillation.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+- Gradient accumulation steps: 4
+- Image resolution: 768
+- Dataset size - 7k images
+- Mixed-precision: fp16
+
+We were able to produce image quality close to the images produced by the original model, with almost 40% fewer parameters and the sample results below speak for themselves:
+
+![image](assets/distill_sd/Picture5.png)
+
+The code for fine-tuning the base models can be found [here](https://github.com/segmind/distill-sd/blob/master/checkpoint_training.py)
```
Comment (sayakpaul) [inline sd_distillation.md:51]:
```suggestion

The code for fine-tuning the base models can be found [here](https://github.com/segmind/distill-sd/blob/master/checkpoint_training.py).
```

File: sd_distillation.md (hunk 8 above)
Comment (sayakpaul) [reply sd_distillation.md:33]: Before coming to the speedup part, I'd maybe show the readers how to use the checkpoints  in `diffusers` to generate images. I'd also briefly discuss the image quality and potential limitations in that section.

File: sd_distillation.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+## LoRA Training
+
+One of the advantages of LoRA training on a distilled model is faster training. Below are some of the images of the first LoRA we trained on the distilled model on some abstract concepts. The code for the LoRA training can be found [here](https://github.com/segmind/distill-sd/blob/master/lora_training.py).
+
+![image](assets/distill_sd/Picture6.png)
+
+## Open-source collaboration
+
+We invite the open-source community to help us improve and achieve wider adoption of these distilled SD models. Users can join our [Discord](https://discord.gg/s6E6eHJk) server, where we will be announcing the latest updates to these models, releasing more checkpoints and some exciting new LoRAs. And if you like our work, please give us a star on our [Github](https://github.com/segmind/distill-sd).
```
Comment (sayakpaul) [inline sd_distillation.md:61]: Maybe this section could be called "Conclusion"?

**Review by sayakpaul — COMMENTED** [review]
Nice start! I love the content and how they have been put together. 

Left a couple of comments.

File: sd_distillation.md (hunk 3 above)
Comment (harishprabhala) [reply sd_distillation.md:8]: sure

File: sd_distillation.md (hunk 5 above)
Comment (harishprabhala) [reply sd_distillation.md:14]: will do

**Comment by kashif** [comment]
@Warlord-K you can remove the images from the assets here as I have added them to the HF spaces here:

https://huggingface.co/datasets/huggingface/documentation-images/discussions/148

File: sd_distillation.md (hunk 1 above)
Comment (Warlord-K) [reply sd_distillation.md]: Changed the Assets as instructed by @kashif

File: sd_distillation.md (hunk 4 above)
Comment (Warlord-K) [reply sd_distillation.md]: Changed

File: sd_distillation.md (hunk 6 above)
Comment (Warlord-K) [reply sd_distillation.md]: Changed

File: sd_distillation.md (hunk 7 above)
Comment (Warlord-K) [reply sd_distillation.md]: Added

File: sd_distillation.md (hunk 8 above)
Comment (Warlord-K) [reply sd_distillation.md]: I have added both these sections, Please take a look.

File: sd_distillation.md (hunk 10 above)
Comment (Warlord-K) [reply sd_distillation.md]: Changed

**Comment by Warlord-K** [comment]
> @Warlord-K you can remove the images from the assets here as I have added them to the HF spaces here:
> 
> https://huggingface.co/datasets/huggingface/documentation-images/discussions/148
> 
> and kindly change the links to the images to the for example:
> 
> ```
> picture 1
> ```

I have changed the links in this format, Please merge the PR in the datasets, so that the images get shown

**Comment by kashif** [comment]
@Warlord-K also remove all the pngs from this PR

File: sd_distillation.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
+---
+title: "Open-sourcing Knowledge Distillation Code and Weights of SD-Small and SD-Tiny"
+thumbnail: /blog/assets/distill_sd/Picture1.png
```
Comment (kashif) [inline sd_distillation.md:3]:
```suggestion
thumbnail: /blog/assets/distill_sd/thumbnail.png

```

File: _blog.yml [hunk 12]
```diff
@@ -2548,3 +2548,13 @@
     - stable-diffusion
     - stable-diffusion-xl
     - diffusers
+
+- local: sd_distillation
+  title: "Open-sourcing Knowledge Distillation Code and Weights of SD-Small and SD-Tiny"
+  author: harishsegmind
+  thumbnail: /blog/assets/distill_sd/Picture1.png
```
Comment (kashif) [inline _blog.yml:2555]:
```suggestion
  thumbnail: /blog/assets/distill_sd/thumbnail.png
```

File: sd_distillation.md (hunk 2 above)
Comment (Warlord-K) [reply sd_distillation.md]: Added

File: _blog.yml [hunk 13]
```diff
@@ -2548,3 +2548,13 @@
...
+  date: August 1, 2023
+  tags:
+    - stable-diffusion
+    - distillation
```
Comment (kashif) [inline _blog.yml:2559]:
```suggestion
```

File: _blog.yml [hunk 14]
```diff
@@ -2548,3 +2548,13 @@
     - stable-diffusion
     - stable-diffusion-xl
     - diffusers
+
+- local: sd_distillation
+  title: "Open-sourcing Knowledge Distillation Code and Weights of SD-Small and SD-Tiny"
+  author: harishsegmind
```
Comment (kashif) [inline _blog.yml:2554]:
```suggestion
  author: harishsegmind
  guest: true
```

Comment (Warlord-K) [reply _blog.yml:2554]: Done

File: _blog.yml (hunk 13 above)
Comment (Warlord-K) [reply _blog.yml]: Done

File: _blog.yml (hunk 12 above)
Comment (Warlord-K) [reply _blog.yml]: Done

File: sd_distillation.md (hunk 11 above)
Comment (Warlord-K) [reply sd_distillation.md:3]: Done

File: sd_distillation.md [hunk 15]
```diff
@@ -0,0 +1,106 @@
+---
+title: "Open-sourcing Knowledge Distillation Code and Weights of SD-Small and SD-Tiny"
+thumbnail: /blog/assets/distill_sd/Picture1.png
+authors:
+- user: harishsegmind
+- user: Warlord-K
+- user: Gothos
```
[the remark this answers is not in the document]
Comment (Warlord-K) [reply sd_distillation.md]: Done

File: sd_distillation.md [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+- user: Warlord-K
+  guest: true
+- user: Gothos
+  guest: true
+---
+
+<h1> Open-sourcing Knowledge Distillation Code and Weights of SD-Small and SD-Tiny </h1>
+
+<p align="center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/distill_sd/thumbnail.png" width=500>
```
Comment (kashif) [inline sd_distillation.md:16]:
```suggestion
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/distill_sd/Picture1.png" width=500>

```

**Review by kashif — APPROVED** [review]
LGTM!

File: sd_distillation.md [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+
+<h1> Open-sourcing Knowledge Distillation Code and Weights of SD-Small and SD-Tiny </h1>
+
+<p align="center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/distill_sd/Picture1.png" width=500>
+
+</p>
+
+
+The past few months have seen a cambrian explosion in larger and larger language models such as GPT-4, Falcon 40B, LLaMa-2 70B etc. The world is zooming towards building large, intelligent and generalized models, increasing the complexity and compute capacities required to run such models. On the imaging side too, Stable diffusion has progressively moved towards larger and larger image generation models such as SD2.1 and SDXL, to deliver highly versatile, state-of-the-art image generations with just a single model, similar to what MidJourney has been doing for the past couple of years. But can we make the models smaller, faster and more specialized, that can eventually run on your everyday electronic devices and even power smart home appliances?
```
Comment (patrickvonplaten) [inline sd_distillation.md:21]: I'd actually argue that the size of LLMs has stagnated a bit. E.g. GPT-3.5 was with 175b parameters much larger then Falcon or Llama

Comment (patrickvonplaten) [inline sd_distillation.md:21]:
```suggestion
The past few months have seen larger and more-performant open-source language models such as Falcon 40B, LLaMa-2 70B, Falcon 40B, MPT 30B etc. The world is zooming towards building large, intelligent and generalized models, increasing the complexity and compute capacities required to run such models. On the imaging side too, Stable diffusion has progressively moved towards larger and larger image generation models such as SD2.1 and SDXL, to deliver highly versatile, state-of-the-art image generations with just a single model, similar to what MidJourney has been doing for the past couple of years. But can we make the models smaller, faster and more specialized, that can eventually run on your everyday electronic devices and even power smart home appliances?
```

Would maybe phrase it slightly more towards open-source and a tad less flashy

Comment (harishprabhala) [reply sd_distillation.md:21]: So, would you like us to redo the line that states LLMs are getting larger?

File: sd_distillation.md [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+The Model can be used using the DiffusionPipeline from [🧨 diffusers](https://github.com/huggingface/diffusers)
+
+```python
+
+from diffusers import DiffusionPipeline
+import torch
+
+pipeline = DiffusionPipeline.from_pretrained("segmind/small-sd", torch_dtype=torch.float16)
+prompt = "Portrait of a pretty girl"
+negative_prompt = "(deformed iris, deformed pupils, semi-realistic, cgi, 3d, render, sketch, cartoon, drawing, anime:1.4), text, close up, cropped, out of frame, worst quality, low quality, jpeg artifacts, ugly, duplicate, morbid, mutilated, extra fingers, mutated hands, poorly drawn hands, poorly drawn face, mutation, deformed, blurry, dehydrated, bad anatomy, bad proportions, extra limbs, cloned face, disfigured, gross proportions, malformed limbs, missing arms, missing legs, extra arms, extra legs, fused fingers, too many fingers, long neck"
```
Comment (patrickvonplaten) [inline sd_distillation.md:60]:
```suggestion
negative_prompt = "deformed iris, deformed pupils, semi-realistic, cgi, 3d, render, sketch, cartoon, drawing, anime, text, close up, cropped, out of frame, worst quality, low quality, jpeg artifacts, ugly, duplicate, morbid, mutilated, extra fingers, mutated hands, poorly drawn hands, poorly drawn face, mutation, deformed, blurry, dehydrated, bad anatomy, bad proportions, extra limbs, cloned face, disfigured, gross proportions, malformed limbs, missing arms, missing legs, extra arms, extra legs, fused fingers, too many fingers, long neck"
```

the prompt weighting doesn't work out of the box with `diffusers` I'm afraid

File: sd_distillation.md [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+image = pipeline(prompt, negative_prompt = negative_prompt).images[0]
```
Comment (patrickvonplaten) [inline sd_distillation.md:61]:
```suggestion
image = pipeline(prompt, negative_prompt=negative_prompt).images[0]

```

**Review by patrickvonplaten — APPROVED** [review]
Looks nice left some smaller comments for potential improvement

File: sd_distillation.md (hunk 17 above)
Comment (harishprabhala) [reply sd_distillation.md:21]: Distilled models are not more specialized per se but given the fewer parameter count than the original models, they may not  be the best generalizable models. But i could be wrong, need to test more :)

Comment (harishprabhala) [reply sd_distillation.md:21]: "In recent times, the AI community has witnessed a remarkable surge in the development of larger and more performant language models, such as Falcon 40B, LLaMa-2 70B, Falcon 40B, MPT 30B, and in the imaging domain with models like SD2.1 and SDXL. These advancements have undoubtedly pushed the boundaries of what AI can achieve, enabling highly versatile and state-of-the-art image generation and language understanding capabilities. However, as we marvel at the power and complexity of these models, it is essential to recognize a growing need to make AI models smaller, efficient, and more accessible, particularly by open-sourcing them."

Is this better @patrickvonplaten ?

**Pull request merged**